modAion/src/org/aion/zero/api/MonetaryCalculator.java

```diff
@@ -0,0 +1,138 @@
+     */
+    public static MonetaryCalculator init(
+            BigInteger initialSupply,
+            int interestBasePoint,
+            long blockAnnum,
+            long startingBlockNum) {
+
+        if (Holder.inst == null) {
+            Holder.setInst(
+                    new MonetaryCalculator(
+                            initialSupply, interestBasePoint, blockAnnum, startingBlockNum));
+        }
+
+        return Holder.inst;
+    }
+
+    private MonetaryCalculator(
+            final BigInteger initialSupply,
+            final int interestBasePoint,
+            final long blockAnnum,
+            final long startingBlockNum) {
+
+        if ((initialSupply == null || initialSupply.signum() != 1)
+                || interestBasePoint < 1
+                || blockAnnum < 1
+                || startingBlockNum < 0) {
+            String s =
+                    "InitialSupply: "
+                            + Objects.requireNonNull(initialSupply).toString()
+                            + " InterestBasePoint: "
+                            + interestBasePoint
+                            + " BlockAnnum: "
+                            + blockAnnum
+                            + " StartingBlockNum:"
+                            + startingBlockNum;
+            throw new IllegalArgumentException(s);
+        }
+
+        INITIALSUPPLY = initialSupply;
+        BLOCKANNUM = blockAnnum;
+        STARTINGBLOCKNUM = startingBlockNum;
+        INFLATIONRATE = BigInteger.valueOf(10000 + interestBasePoint);
+    }
+
+    /**
+     * Given the block number desire to calculate the reward. The basic term will be 1 then
+     * increased by the annum of blocks. Should call {@link MonetaryCalculator.init()} before use
+     * it.
+     *
+     * @param blockNum The block number for calculating the reward.
+     * @return The current token reward.
+     * @exception IllegalArgumentException The blockNum should not less or equal to the targeting
+     *     block number.
+     * @exception Exception the method should be called after the Holder initialized.
+     */
+    public static BigInteger getCurrentReward(final long blockNum) {
+        if (Holder.inst == null) {
+            throw new NullPointerException(
+                    "No MonetaryCalculator Holder instance, please initial it before the call.");
+        }
+
+        if (blockNum <= STARTINGBLOCKNUM) {
+            throw new IllegalArgumentException("Invalid blockNum:" + blockNum);
+        }
+
+        long term = (blockNum - STARTINGBLOCKNUM - 1) / BLOCKANNUM + 1;
+
+        if (term != CURRENTTERM) {
+            CURRENTREWARD = calculateCompound(term);
+            CURRENTTERM = term;
+        }
+
+        return CURRENTREWARD;
+    }
+
+    private static BigInteger calculateCompound(long term) {
+        BigInteger compound = DEVIDER.multiply(INITIALSUPPLY);
+        BigInteger preCompound = compound;
+        for (long i = 0; i < term; i++) {
```

Comment:
If possible, I would like to bound this function, and assume that any block that violates this bound fails. From there we can generate a LUT for the next 1024 years (so we get O(1)). The reason being that if we were ever to introduce an attack vector that ran this calculation without verifying the block height, a user could reasonably exploit the linear complexity of this calculation to stall the thread executing this.

I will leave final decision up to the implementers. But I would just be wary that this does introduce an avenue of attack.

Reply:
Also note: I looked at where rewards are being calculated, I don't believe that there is _currently_ any way to exploit this.

Reply:
updated in 
https://github.com/aionnetwork/aion/pull/852/commits/2f83196e6d8f13cea399a351c8cf50fb3a77f803